File: stem-game-backend/stem-game-backend/src/models/Topic.js
const { DataTypes } = require('sequelize');
const { sequelize } = require('../config/database');

const Topic = sequelize.define('Topic', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  islandId: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: 'islands',
      key: 'id',
    },
    onDelete: 'CASCADE',
    comment: 'Foreign key to the island this topic belongs to',
  },
  code: {
    type: DataTypes.STRING(50),
    allowNull: false,
    unique: true,
    comment: 'Unique code for the topic (e.g., "topic_shadows")',
  },
  name: {
    type: DataTypes.STRING(100),
    allowNull: false,
    comment: 'Display name of the topic (e.g., "Shadows & Light")',
  },
  description: {
    type: DataTypes.TEXT,
    allowNull: true,
    comment: 'Description of what students will learn in this topic',
  },
  learningObjectives: {
    type: DataTypes.JSONB,
    allowNull: true,
    defaultValue: [],
    comment: 'Array of learning objectives for this topic',
  },
  orderIndex: {
    type: DataTypes.INTEGER,
    allowNull: false,
    validate: {
      min: 1,
    },
    comment: 'Display order within the island',
  },
  iconUrl: {
    type: DataTypes.STRING(255),
    allowNull: true,
    comment: 'URL to the topic icon/image',
  },
  difficultyLevel: {
    type: DataTypes.ENUM('beginner', 'intermediate', 'advanced'),
    allowNull: false,
    defaultValue: 'beginner',
    comment: 'Difficulty level of the topic',
  },
  levelCount: {
    type: DataTypes.INTEGER,
    allowNull: false,
    defaultValue: 8,
    validate: {
      min: 1,
    },
    comment: 'Number of levels in this topic',
  },
}, {
  tableName: 'topics',
  timestamps: true,
  indexes: [
    {
      unique: true,
      fields: ['code'],
    },
    {
      fields: ['islandId', 'orderIndex'],
    },
    {
      fields: ['difficultyLevel'],
    },
  ],
});

module.exports = Topic;

File: stem-game-flutter/stem-game-flutter/lib/features/world/models/topic.dart
import 'package:equatable/equatable.dart';
import 'island.dart';

class Topic extends Equatable {
  final String id;
  final String islandId;
  final String code;
  final String name;
  final String description;
  final List<String> learningObjectives;
  final int orderIndex;
  final String? iconUrl;
  final String difficultyLevel;
  final int levelCount;
  final bool? isUnlocked;
  final IslandProgress? userProgress;

  const Topic({
    required this.id,
    required this.islandId,
    required this.code,
    required this.name,
    required this.description,
    required this.learningObjectives,
    required this.orderIndex,
    this.iconUrl,
    required this.difficultyLevel,
    required this.levelCount,
    this.isUnlocked,
    this.userProgress,
  });

  factory Topic.fromJson(Map<String, dynamic> json) {
    return Topic(
      id: json['id'] as String? ?? '',
      islandId: json['islandId'] as String? ?? '',
      code: json['code'] as String? ?? '',
      name: json['name'] as String? ?? 'Unknown Topic',
      description: json['description'] as String? ?? '',
      learningObjectives: (json['learningObjectives'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      orderIndex: json['orderIndex'] as int? ?? 0,
      iconUrl: json['iconUrl'] as String?,
      difficultyLevel: json['difficultyLevel'] as String? ?? 'beginner',
      levelCount: json['levelCount'] as int? ?? 0,
      isUnlocked: json['isUnlocked'] as bool?,
      userProgress: json['userProgress'] != null
          ? IslandProgress.fromJson(json['userProgress'] as Map<String, dynamic>)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'islandId': islandId,
      'code': code,
      'name': name,
      'description': description,
      'learningObjectives': learningObjectives,
      'orderIndex': orderIndex,
      if (iconUrl != null) 'iconUrl': iconUrl,
      'difficultyLevel': difficultyLevel,
      'levelCount': levelCount,
      if (isUnlocked != null) 'isUnlocked': isUnlocked,
      if (userProgress != null) 'userProgress': userProgress!.toJson(),
    };
  }

  @override
  List<Object?> get props => [
        id,
        islandId,
        code,
        name,
        description,
        learningObjectives,
        orderIndex,
        iconUrl,
        difficultyLevel,
        levelCount,
        isUnlocked,
        userProgress,
      ];
}
